From 97cf0413cdb40444cc4f1e521dc6e7b912a8a61e Mon Sep 17 00:00:00 2001
From: Anilkumar Kolli <akolli@codeaurora.org>
Date: Mon, 14 Sep 2020 16:41:47 +0530
Subject: [PATCH] ath11k: add full monitor support for QCN9000

1) add hw specific callback to differentiate legacy and full
   monitor mode support.
2) config cmd HTT_H2T_MSG_TYPE_RX_FULL_MONITOR_MODE is
   sent to configure full monitor mode.
3) mon_dst_ring ring is read in sw_mon_ring decriptor format
   instead of reo_entrance_ring descriptor format.
4) Mon_dst_ring is reaped first and then mon_status ring.
5) Status is available for all entries in mon_dst ring, this
   is achieved by using HW lock-stepping.
6) Interrupt received for each end of ppdu in dst ring, hw also
   copies status for same PPDU.

Signed-off-by: Anilkumar Kolli <akolli@codeaurora.org>
---
 drivers/net/wireless/ath/ath11k/ahb.c      |   1 +
 drivers/net/wireless/ath/ath11k/core.c     |   3 +
 drivers/net/wireless/ath/ath11k/dp.c       |   7 +-
 drivers/net/wireless/ath/ath11k/dp.h       |  99 +++++++
 drivers/net/wireless/ath/ath11k/dp_rx.c    | 425 +++++++++++++++++++++++++++--
 drivers/net/wireless/ath/ath11k/dp_rx.h    |   4 -
 drivers/net/wireless/ath/ath11k/dp_tx.c    |  39 +++
 drivers/net/wireless/ath/ath11k/dp_tx.h    |   2 +
 drivers/net/wireless/ath/ath11k/hal_desc.h |  19 ++
 drivers/net/wireless/ath/ath11k/hal_rx.c   |  47 ++++
 drivers/net/wireless/ath/ath11k/hal_rx.h   |  17 ++
 drivers/net/wireless/ath/ath11k/hw.c       |  16 ++
 drivers/net/wireless/ath/ath11k/hw.h       |  11 +
 drivers/net/wireless/ath/ath11k/mac.c      |  11 +
 14 files changed, 677 insertions(+), 24 deletions(-)

Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/ahb.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/ahb.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/ahb.c
@@ -12,6 +12,7 @@
 #include "debug.h"
 #include "hif.h"
 #include "qmi.h"
+#include "dp_rx.h"
 #include <linux/remoteproc.h>
 
 static const u8 ath11k_tx_ring_mask[ATH11K_EXT_IRQ_GRP_NUM_MAX] = {
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/core.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/core.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/core.c
@@ -44,6 +44,7 @@ static const struct ath11k_hw_params ath
 		.max_radios = 3,
 		.bdf_addr = 0x4B0C0000,
 		.mon_status_intr_threshold = 0,
+		.full_monitor_mode = false,
 		.bdf_sz = ATH11K_QMI_BDF_MAX_SIZE,
 		.caldb_sz = ATH11K_QMI_CALDB_SIZE_IPQ807X,
 		.hal_desc_sz = sizeof(struct hal_rx_desc_ipq8074),
@@ -66,6 +67,7 @@ static const struct ath11k_hw_params ath
 		.max_radios = 2,
 		.bdf_addr = 0x4ABC0000,
 		.mon_status_intr_threshold = 0,
+		.full_monitor_mode = false,
 		.bdf_sz = ATH11K_QMI_BDF_MAX_SIZE,
 		.caldb_sz = ATH11K_QMI_CALDB_SIZE_IPQ807X,
 		.hal_desc_sz = sizeof(struct hal_rx_desc_ipq8074),
@@ -87,6 +89,7 @@ static const struct ath11k_hw_params ath
 		.max_radios = 1,
 		.bdf_addr = 0x54800000,
 		.mon_status_intr_threshold = 4,
+		.full_monitor_mode = true,
 		.bdf_sz = ATH11K_QMI_HOST_FWADDR_SZ_QCN9000,
 		.caldb_sz = ATH11K_QMI_CALDB_SIZE_QCN9000,
 		.hal_desc_sz = sizeof(struct hal_rx_desc_qcn9000),
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/dp.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp.c
@@ -800,9 +800,9 @@ int ath11k_dp_service_srng(struct ath11k
 		for (i = 0; i <  ab->num_radios; i++) {
 			if (ab->rx_mon_status_ring_mask[grp_id] & BIT(i)) {
 				work_done =
-				ath11k_dp_rx_process_mon_rings(ab,
-							       i, napi,
-							       budget);
+				ab->hw_params.hw_ops->rx_process_mon_rings(ab,
+									   i, napi,
+									   budget);
 				budget -= work_done;
 				tot_work_done += work_done;
 			}
@@ -1016,6 +1016,7 @@ int ath11k_dp_alloc(struct ath11k_base *
 
 	INIT_LIST_HEAD(&dp->reo_cmd_list);
 	INIT_LIST_HEAD(&dp->reo_cmd_cache_flush_list);
+	INIT_LIST_HEAD(&dp->dp_full_mon_mpdu_list);
 	spin_lock_init(&dp->reo_cmd_lock);
 
 	dp->reo_cmd_cache_flush_count = 0;
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp.h
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/dp.h
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp.h
@@ -84,6 +84,19 @@ struct dp_tx_ring {
 	int tx_status_tail;
 };
 
+enum dp_mon_reap_status {
+	/* PPDU id matches in dst ring and status ring */
+	DP_MON_STATUS_MATCH,
+	/* status ring dma is not done */
+	DP_MON_STATUS_NO_DMA,
+	/* status ring is lagging, reap status ring */
+	DP_MON_STATUS_LAG,
+	/* status ring is leading, reap dst ring and drop */
+	DP_MON_STATUS_LEAD,
+	/* replinish monitor status ring */
+	DP_MON_STATUS_REPLINISH,
+};
+
 struct ath11k_pdev_mon_stats {
 	u32 status_ppdu_state;
 	u32 status_ppdu_start;
@@ -101,6 +114,12 @@ struct ath11k_pdev_mon_stats {
 	u32 dest_mon_not_reaped;
 };
 
+struct dp_full_mon_mpdu {
+	struct list_head list;
+	struct sk_buff *head;
+	struct sk_buff *tail;
+};
+
 struct dp_link_desc_bank {
 	void *vaddr_unaligned;
 	void *vaddr;
@@ -139,7 +158,12 @@ struct ath11k_mon_data {
 	u32 mon_last_buf_cookie;
 	u64 mon_last_linkdesc_paddr;
 	u16 chan_noise_floor;
+	bool hold_mon_dst_ring;
+	u32 mon_status_buf_state;
+	dma_addr_t mon_status_paddr;
 
+	struct dp_full_mon_mpdu *mon_mpdu;
+	struct hal_sw_mon_ring_entries sw_mon_entries;
 	struct ath11k_pdev_mon_stats rx_mon_stats;
 	/* lock for monitor data */
 	spinlock_t mon_lock;
@@ -235,6 +259,7 @@ struct ath11k_dp {
 	struct hal_wbm_idle_scatter_list scatter_list[DP_IDLE_SCATTER_BUFS_MAX];
 	struct list_head reo_cmd_list;
 	struct list_head reo_cmd_cache_flush_list;
+	struct list_head dp_full_mon_mpdu_list;
 	u32 reo_cmd_cache_flush_count;
 	/**
 	 * protects access to below fields,
@@ -280,6 +305,7 @@ enum htt_h2t_msg_type {
 	HTT_H2T_MSG_TYPE_RX_RING_SELECTION_CFG	= 0xc,
 	HTT_H2T_MSG_TYPE_EXT_STATS_CFG		= 0x10,
 	HTT_H2T_MSG_TYPE_PPDU_STATS_CFG		= 0x11,
+	HTT_H2T_MSG_TYPE_RX_FULL_MONITOR_MODE	= 0x17,
 };
 
 #define HTT_VER_REQ_INFO_MSG_ID		GENMASK(7, 0)
@@ -983,6 +1009,79 @@ struct htt_rx_ring_tlv_filter {
 	u32 pkt_filter_flags3; /* DATA */
 };
 
+/**
+ * @brief Host-->target HTT RX Full monitor mode register configuration message
+ * @details
+ * The host will send this Full monitor mode register configuration message.
+ * This message can be sent per SOC or per PDEV which is differentiated
+ * by pdev id values.
+ *
+ *       |31                            16|15  11|10   8|7      3|2|1|0|
+ *       |-------------------------------------------------------------|
+ *       |             reserved           |   pdev_id   |  MSG_TYPE    |
+ *       |-------------------------------------------------------------|
+ *       |                      reserved         |Release Ring   |N|Z|E|
+ *       |-------------------------------------------------------------|
+ *
+ * where E  is 1-bit full monitor mode enable/disable.
+ *       Z  is 1-bit additional descriptor for zero mpdu enable/disable
+ *       N  is 1-bit additional descriptor for non zero mdpu enable/disable
+ *
+ * The following field definitions describe the format of the full monitor
+ * mode configuration message sent from the host to target for each pdev.
+ *
+ * Header fields:
+ *  dword0 - b'7:0   - msg_type: This will be set to
+ *                     HTT_H2T_MSG_TYPE_RX_FULL_MONITOR_MODE.
+ *           b'15:8  - pdev_id:  0 indicates msg is for all LMAC rings, i.e. soc
+ *                     1, 2, 3 indicates pdev_id 0,1,2 and the msg is for the
+ *                     specified pdev's LMAC ring.
+ *           b'31:16 - reserved : Reserved for future use.
+ *  dword1 - b'0     - full_monitor_mode enable: This indicates that the full
+ *                     monitor mode rxdma register is to be enabled or disabled.
+ *           b'1     - addnl_descs_zero_mpdus_end: This indicates that the
+ *                     additional descriptors at ppdu end for zero mpdus
+ *                     enabled or disabled.
+ *           b'2     - addnl_descs_non_zero_mpdus_end: This indicates that the
+ *                     additional descriptors at ppdu end for non zero mpdus
+ *                     enabled or disabled.
+ *           b'10:3  - release_ring: This indicates the destination ring
+ *                     selection for the descriptor at the end of PPDU
+ *                     0 - REO ring select
+ *                     1 - FW  ring select
+ *                     2 - SW  ring select
+ *                     3 - Release ring select
+ *                     Refer to htt_rx_full_mon_release_ring.
+ *           b'31:11  - reserved for future use
+ */
+
+#define HTT_RX_FULL_MON_MODE_CFG_CMD_INFO0_MSG_TYPE	GENMASK(7, 0)
+#define HTT_RX_FULL_MON_MODE_CFG_CMD_INFO0_PDEV_ID	GENMASK(15, 8)
+
+#define HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_ENABLE			BIT(0)
+#define HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_ZERO_MPDUS_END		BIT(1)
+#define HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_NON_ZERO_MPDUS_END	BIT(2)
+#define HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_RELEASE_RING		GENMASK(10, 3)
+
+/**
+ * Enumeration for full monitor mode destination ring select
+ * 0 - REO destination ring select
+ * 1 - FW destination ring select
+ * 2 - SW destination ring select
+ * 3 - Release destination ring select
+ */
+enum htt_rx_full_mon_release_ring {
+	HTT_RX_MON_RING_REO,
+	HTT_RX_MON_RING_FW,
+	HTT_RX_MON_RING_SW,
+	HTT_RX_MON_RING_RELEASE,
+};
+
+struct htt_rx_full_monitor_mode_cfg_cmd {
+	u32 info0;
+	u32 cfg;
+} __packed;
+
 /* HTT message target->host */
 
 enum htt_t2h_msg_type {
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -3186,12 +3186,48 @@ fail_desc_get:
 	return req_entries - num_remain;
 }
 
+#define DP_RX_FULL_MON_PPDU_ID_WRAP 32535
+
+static void
+ath11k_dp_rx_mon_update_status_buf_state(struct ath11k_mon_data *pmon,
+					 struct hal_tlv_hdr *tlv)
+{
+	u16 ppdu_id_diff, ppdu_id, tlv_len;
+	u8 *ptr = tlv;
+
+	tlv_len = FIELD_GET(HAL_TLV_HDR_LEN, tlv->tl);
+	ptr += sizeof(*tlv) + tlv_len;
+
+	tlv = (struct hal_tlv_hdr *)ptr;
+
+	if (FIELD_GET(HAL_TLV_HDR_TAG, tlv->tl) == HAL_RX_PPDU_START) {
+		struct hal_rx_ppdu_start *ppdu_start =
+			(struct hal_rx_ppdu_start *)(ptr + sizeof(*tlv));
+
+		ppdu_id = FIELD_GET(HAL_RX_PPDU_START_INFO0_PPDU_ID,
+				   __le32_to_cpu(ppdu_start->info0));
+	}
+
+	if (pmon->sw_mon_entries.ppdu_id < ppdu_id) {
+		pmon->mon_status_buf_state = DP_MON_STATUS_LEAD;
+		ppdu_id_diff = ppdu_id - pmon->sw_mon_entries.ppdu_id;
+		if (ppdu_id_diff > DP_RX_FULL_MON_PPDU_ID_WRAP)
+			pmon->mon_status_buf_state = DP_MON_STATUS_LAG;
+	} else if (pmon->sw_mon_entries.ppdu_id > ppdu_id) {
+		pmon->mon_status_buf_state = DP_MON_STATUS_LAG;
+		ppdu_id_diff = pmon->sw_mon_entries.ppdu_id - ppdu_id;
+		if (ppdu_id_diff > DP_RX_FULL_MON_PPDU_ID_WRAP)
+			pmon->mon_status_buf_state = DP_MON_STATUS_LEAD;
+	}
+}
+
 static int ath11k_dp_rx_reap_mon_status_ring(struct ath11k_base *ab, int mac_id,
 					     int *budget, struct sk_buff_head *skb_list)
 {
 	struct ath11k *ar = ab->pdevs[mac_id].ar;
 	struct ath11k_pdev_dp *dp = &ar->dp;
 	struct dp_rxdma_ring *rx_ring = &dp->rx_mon_status_refill_ring;
+	struct ath11k_mon_data *pmon = (struct ath11k_mon_data *)&dp->mon_data;
 	struct hal_srng *srng;
 	void *rx_mon_status_desc;
 	struct sk_buff *skb;
@@ -3212,8 +3248,10 @@ static int ath11k_dp_rx_reap_mon_status_
 		*budget -= 1;
 		rx_mon_status_desc =
 			ath11k_hal_srng_src_peek(ab, srng);
-		if (!rx_mon_status_desc)
+		if (!rx_mon_status_desc) {
+			pmon->mon_status_buf_state = DP_MON_STATUS_REPLINISH;
 			break;
+		}
 
 		ath11k_hal_rx_buf_addr_info_get(rx_mon_status_desc, &paddr,
 						&cookie, &rbm);
@@ -3226,6 +3264,7 @@ static int ath11k_dp_rx_reap_mon_status_
 				ath11k_warn(ab, "rx monitor status with invalid buf_id %d\n",
 					    buf_id);
 				spin_unlock_bh(&rx_ring->idr_lock);
+				pmon->mon_status_buf_state = DP_MON_STATUS_REPLINISH;
 				goto move_next;
 			}
 
@@ -3249,15 +3288,23 @@ static int ath11k_dp_rx_reap_mon_status_
 					    FIELD_GET(HAL_TLV_HDR_TAG,
 						      tlv->tl));
 				dev_kfree_skb_any(skb);
+				pmon->mon_status_buf_state = DP_MON_STATUS_NO_DMA;
 				goto move_next;
 			}
 
+			if (ab->hw_params.full_monitor_mode) {
+				ath11k_dp_rx_mon_update_status_buf_state(pmon, tlv);
+				if (paddr == pmon->mon_status_paddr)
+					pmon->mon_status_buf_state = DP_MON_STATUS_MATCH;
+			}
+
 			__skb_queue_tail(skb_list, skb);
+		} else {
+			pmon->mon_status_buf_state = DP_MON_STATUS_REPLINISH;
 		}
 move_next:
 		skb = ath11k_dp_rx_alloc_mon_status_buf(ab, rx_ring,
 							&buf_id, GFP_ATOMIC);
-
 		if (!skb) {
 			ath11k_hal_rx_buf_addr_info_set(rx_mon_status_desc, 0, 0,
 							HAL_RX_BUF_RBM_SW3_BM);
@@ -3271,6 +3318,7 @@ move_next:
 
 		ath11k_hal_rx_buf_addr_info_set(rx_mon_status_desc, rxcb->paddr,
 						cookie, HAL_RX_BUF_RBM_SW3_BM);
+
 		ath11k_hal_srng_src_get_next_entry(ab, srng);
 		num_buffs_reaped++;
 	}
@@ -3384,8 +3432,10 @@ int ath11k_dp_rx_process_mon_status(stru
 		    (pmon->mon_ppdu_status == DP_PPDU_STATUS_START)) {
 			rx_mon_stats->status_ppdu_done++;
 			pmon->mon_ppdu_status = DP_PPDU_STATUS_DONE;
-			ath11k_dp_rx_mon_dest_process(ar, budget, napi);
-			pmon->mon_ppdu_status = DP_PPDU_STATUS_START;
+			if (!ab->hw_params.full_monitor_mode) {
+				ath11k_dp_rx_mon_dest_process(ar, budget, napi);
+				pmon->mon_ppdu_status = DP_PPDU_STATUS_START;
+			}
 		}
 
 		if (ppdu_info->peer_id == HAL_INVALID_PEERID ||
@@ -3426,7 +3476,6 @@ int ath11k_dp_rx_process_mon_status(stru
 
 		spin_unlock_bh(&ab->base_lock);
 		rcu_read_unlock();
-
 		dev_kfree_skb_any(skb);
 	}
 exit:
@@ -5271,12 +5320,10 @@ static int ath11k_dp_rx_mon_deliver(stru
 
 	mon_skb = ath11k_dp_rx_mon_merg_msdus(ar, mac_id, head_msdu,
 					      tail_msdu, rxs);
-
 	if (!mon_skb)
 		goto mon_deliver_fail;
 
 	header = mon_skb;
-
 	rxs->flag = 0;
 	do {
 		skb_next = mon_skb->next;
@@ -5315,11 +5362,11 @@ void ath11k_dp_rx_mon_dest_process(struc
 {
 	struct ath11k_pdev_dp *dp = &ar->dp;
 	struct ath11k_mon_data *pmon = (struct ath11k_mon_data *)&dp->mon_data;
+	struct ath11k_pdev_mon_stats *rx_mon_stats;
 	void *ring_entry;
 	void *mon_dst_srng;
 	u32 ppdu_id;
 	u32 rx_bufs_used;
-	struct ath11k_pdev_mon_stats *rx_mon_stats;
 	u32	 npackets = 0;
 	u32 mpdu_rx_bufs_used;
 
@@ -5381,6 +5428,7 @@ void ath11k_dp_rx_mon_dest_process(struc
 				   rx_mon_stats->dest_mon_stuck);
 			break;
 		}
+
 		if (head_msdu && tail_msdu) {
 			ath11k_dp_rx_mon_deliver(ar, dp->mac_id, head_msdu,
 						 &pmon->mon_ppdu_info,
@@ -5404,15 +5452,6 @@ void ath11k_dp_rx_mon_dest_process(struc
 	}
 }
 
-int ath11k_dp_rx_process_mon_rings(struct ath11k_base *ab, int mac_id,
-				   struct napi_struct *napi, int budget)
-{
-	int ret = 0;
-
-	ret = ath11k_dp_rx_process_mon_status(ab, mac_id, napi, budget);
-	return ret;
-}
-
 static int ath11k_dp_rx_pdev_mon_status_attach(struct ath11k *ar)
 {
 	struct ath11k_pdev_dp *dp = &ar->dp;
@@ -5477,3 +5516,355 @@ int ath11k_dp_rx_pdev_mon_detach(struct
 	ath11k_dp_mon_link_free(ar);
 	return 0;
 }
+
+static u32
+ath11k_dp_rx_full_mon_mpdu_pop(struct ath11k *ar,
+			       void *ring_entry, struct sk_buff **head_msdu,
+			       struct sk_buff **tail_msdu,
+			       struct hal_sw_mon_ring_entries *sw_mon_entries)
+{
+	struct ath11k_pdev_dp *dp = &ar->dp;
+	struct ath11k_mon_data *pmon = (struct ath11k_mon_data *)&dp->mon_data;
+	struct dp_rxdma_ring *rx_ring = &dp->rxdma_mon_buf_ring;
+	struct sk_buff *msdu = NULL, *last = NULL;
+	struct hal_sw_monitor_ring *sw_desc =
+			(struct hal_sw_monitor_ring*)ring_entry;
+	struct hal_rx_msdu_list msdu_list;
+	struct hal_rx_desc *rx_desc;
+	struct ath11k_pdev_mon_stats *rx_mon_stats;
+	struct ath11k_skb_rxcb *rxcb;
+	void *rx_msdu_link_desc;
+	void *p_buf_addr_info, *p_last_buf_addr_info;
+	int buf_id, i = 0;
+	u32 rx_buf_size, rx_pkt_offset, l2_hdr_offset;
+	u32 rx_bufs_used = 0;
+	u32 msdu_cnt = 0;
+	u32 total_len = 0, frag_len = 0, sw_cookie;
+	u16 num_msdus = 0;
+	u8 rxdma_err;
+	bool is_frag;
+	bool drop_mpdu = false;
+
+	rx_mon_stats = &pmon->rx_mon_stats;
+
+	ath11k_hal_rx_sw_mon_ring_buf_paddr_get(ring_entry, sw_mon_entries);
+
+	sw_cookie = sw_mon_entries->mon_dst_sw_cookie;
+	sw_mon_entries->end_of_ppdu = false;
+	sw_mon_entries->drop_ppdu = false;
+	p_last_buf_addr_info = sw_mon_entries->dst_buf_addr_info;
+	msdu_cnt = sw_mon_entries->msdu_cnt;
+
+	sw_mon_entries->end_of_ppdu =
+		FIELD_GET(HAL_SW_MON_RING_INFO0_END_OF_PPDU, sw_desc->info0);
+	if (sw_mon_entries->end_of_ppdu)
+		return rx_bufs_used;
+
+	if (FIELD_GET(HAL_SW_MON_RING_INFO0_RXDMA_PUSH_REASON,
+		      sw_desc->info0) ==
+		      HAL_REO_DEST_RING_PUSH_REASON_ERR_DETECTED) {
+		rxdma_err =
+			FIELD_GET(HAL_SW_MON_RING_INFO0_RXDMA_ERROR_CODE,
+				  sw_desc->info0);
+		if (rxdma_err == HAL_REO_ENTR_RING_RXDMA_ECODE_FLUSH_REQUEST_ERR ||
+		    rxdma_err == HAL_REO_ENTR_RING_RXDMA_ECODE_MPDU_LEN_ERR ||
+		    rxdma_err == HAL_REO_ENTR_RING_RXDMA_ECODE_OVERFLOW_ERR) {
+			pmon->rx_mon_stats.dest_mpdu_drop++;
+			drop_mpdu = true;
+		}
+	}
+
+	do {
+		rx_msdu_link_desc =
+			(void *)pmon->link_desc_banks[sw_cookie].vaddr +
+			(sw_mon_entries->mon_dst_paddr -
+			 pmon->link_desc_banks[sw_cookie].paddr);
+
+		ath11k_hal_rx_msdu_list_get(ar, rx_msdu_link_desc, &msdu_list,
+					    &num_msdus);
+
+		for (i = 0; i < num_msdus; i++) {
+
+			buf_id = FIELD_GET(DP_RXDMA_BUF_COOKIE_BUF_ID,
+					   msdu_list.sw_cookie[i]);
+
+			spin_lock_bh(&rx_ring->idr_lock);
+			msdu = idr_find(&rx_ring->bufs_idr, buf_id);
+			if (!msdu) {
+				ath11k_dbg(ar->ab, ATH11K_DBG_DATA,
+					   "full mon msdu_pop: invalid buf_id %d\n",
+					    buf_id);
+				spin_unlock_bh(&rx_ring->idr_lock);
+				break;
+			}
+			idr_remove(&rx_ring->bufs_idr, buf_id);
+			spin_unlock_bh(&rx_ring->idr_lock);
+
+			rxcb = ATH11K_SKB_RXCB(msdu);
+			if (!rxcb->unmapped) {
+				dma_unmap_single(ar->ab->dev, rxcb->paddr,
+						 msdu->len +
+						 skb_tailroom(msdu),
+						 DMA_FROM_DEVICE);
+				rxcb->unmapped = 1;
+			}
+			if (drop_mpdu) {
+				ath11k_dbg(ar->ab, ATH11K_DBG_DATA,
+					   "full mon: i %d drop msdu %p *ppdu_id %x\n",
+					   i, msdu, sw_mon_entries->ppdu_id);
+				dev_kfree_skb_any(msdu);
+				msdu_cnt--;
+				goto next_msdu;
+			}
+
+			rx_desc = (struct hal_rx_desc *)msdu->data;
+
+			rx_pkt_offset = sizeof(struct hal_rx_desc);
+			l2_hdr_offset = ath11k_dp_rx_h_msdu_end_l3pad(ar->ab, rx_desc);
+
+			ath11k_dp_mon_get_buf_len(&msdu_list.msdu_info[i],
+						  &is_frag, &total_len,
+						  &frag_len, &msdu_cnt);
+
+			rx_buf_size = rx_pkt_offset + l2_hdr_offset + frag_len;
+
+			ath11k_dp_pkt_set_pktlen(msdu, rx_buf_size);
+
+			if (!(*head_msdu))
+				*head_msdu = msdu;
+			else if (last)
+				last->next = msdu;
+
+			last = msdu;
+next_msdu:
+			rx_bufs_used++;
+		}
+
+		ath11k_dp_rx_mon_next_link_desc_get(rx_msdu_link_desc,
+						    &sw_mon_entries->mon_dst_paddr,
+						    &sw_mon_entries->mon_dst_sw_cookie,
+						    &p_buf_addr_info);
+
+		if (ath11k_dp_rx_monitor_link_desc_return(ar,
+							  p_last_buf_addr_info,
+							  dp->mac_id))
+			ath11k_dbg(ar->ab, ATH11K_DBG_DATA,
+				   "full mon: dp_rx_monitor_link_desc_return failed\n");
+
+		p_last_buf_addr_info = p_buf_addr_info;
+
+	} while (sw_mon_entries->mon_dst_paddr && msdu_cnt);
+
+	if (last)
+		last->next = NULL;
+
+	*tail_msdu = msdu;
+
+	return rx_bufs_used;
+}
+
+static int ath11k_dp_rx_full_mon_prepare_mpdu(struct ath11k_dp *dp,
+					       struct dp_full_mon_mpdu *mon_mpdu,
+					       struct sk_buff *head,
+					       struct sk_buff *tail)
+{
+	mon_mpdu = kzalloc(sizeof(*mon_mpdu), GFP_ATOMIC);
+	if (!mon_mpdu)
+		return -ENOMEM;
+
+	list_add_tail(&mon_mpdu->list, &dp->dp_full_mon_mpdu_list);
+	mon_mpdu->head = head;
+	mon_mpdu->tail = tail;
+
+	return 0;
+}
+
+static void ath11k_dp_rx_full_mon_drop_ppdu(struct ath11k_dp *dp,
+					    struct dp_full_mon_mpdu *mon_mpdu)
+{
+	struct dp_full_mon_mpdu *tmp;
+	struct sk_buff *tmp_msdu, *skb_next;
+
+	if (list_empty(&dp->dp_full_mon_mpdu_list))
+		return;
+
+	list_for_each_entry_safe(mon_mpdu, tmp, &dp->dp_full_mon_mpdu_list, list) {
+		list_del(&mon_mpdu->list);
+
+		tmp_msdu = mon_mpdu->head;
+		while (tmp_msdu) {
+			skb_next = tmp_msdu->next;
+			dev_kfree_skb_any(tmp_msdu);
+			tmp_msdu = skb_next;
+		}
+
+		kfree(mon_mpdu);
+	}
+}
+
+static int ath11k_dp_rx_full_mon_deliver_ppdu(struct ath11k *ar,
+					      int mac_id,
+					      struct ath11k_mon_data *pmon,
+					      struct napi_struct *napi)
+{
+	struct ath11k_pdev_mon_stats *rx_mon_stats;
+	struct dp_full_mon_mpdu *tmp;
+	struct dp_full_mon_mpdu *mon_mpdu = pmon->mon_mpdu;
+	struct hal_rx_mon_ppdu_info *ppdu_info;
+	struct sk_buff *head_msdu, *tail_msdu;
+	struct ath11k_base *ab = ar->ab;
+	struct ath11k_dp *dp = &ab->dp;
+	int ret;
+
+	rx_mon_stats = &pmon->rx_mon_stats;
+	ppdu_info = &pmon->mon_ppdu_info;
+
+	list_for_each_entry_safe(mon_mpdu, tmp, &dp->dp_full_mon_mpdu_list, list) {
+		list_del(&mon_mpdu->list);
+		head_msdu = mon_mpdu->head;
+		tail_msdu = mon_mpdu->tail;
+		if (head_msdu && tail_msdu) {
+			ret = ath11k_dp_rx_mon_deliver(ar, mac_id, head_msdu,
+						       ppdu_info,
+						       tail_msdu, napi);
+			rx_mon_stats->dest_mpdu_done++;
+			ath11k_dbg(ar->ab, ATH11K_DBG_DATA, "full mon: deliver ppdu\n");
+		}
+		kfree(mon_mpdu);
+	}
+
+	return ret;
+}
+
+static int
+ath11k_dp_rx_process_full_mon_status_ring(struct ath11k_base *ab, int mac_id,
+					  struct napi_struct *napi, int budget)
+{
+	struct ath11k *ar = ab->pdevs[mac_id].ar;
+	struct ath11k_pdev_dp *dp = &ar->dp;
+	struct ath11k_mon_data *pmon = (struct ath11k_mon_data *)&dp->mon_data;
+	struct hal_sw_mon_ring_entries *sw_mon_entries;
+	int quota = 0, work = 0;
+
+	sw_mon_entries = &pmon->sw_mon_entries;
+
+	while(pmon->hold_mon_dst_ring) {
+		quota = ath11k_dp_rx_process_mon_status(ab, mac_id,
+							napi, 1);
+		if (pmon->mon_status_buf_state == DP_MON_STATUS_MATCH) {
+			ath11k_dp_rx_full_mon_deliver_ppdu(ar, dp->mac_id,
+							   pmon, napi);
+			pmon->hold_mon_dst_ring = false;
+		} else if (!pmon->mon_status_paddr ||
+			   pmon->mon_status_buf_state == DP_MON_STATUS_LEAD) {
+			sw_mon_entries->drop_ppdu = true;
+			pmon->hold_mon_dst_ring = false;
+		}
+
+		if (!quota)
+			break;
+
+		work += quota;
+	}
+
+	if(sw_mon_entries->drop_ppdu)
+		ath11k_dp_rx_full_mon_drop_ppdu(&ab->dp, pmon->mon_mpdu);
+
+	return work;
+}
+
+int ath11k_dp_rx_process_full_monitor(struct ath11k_base *ab, int mac_id,
+				      struct napi_struct *napi, int budget)
+{
+	struct ath11k *ar = ab->pdevs[mac_id].ar;
+	struct ath11k_pdev_dp *dp = &ar->dp;
+	struct ath11k_mon_data *pmon = (struct ath11k_mon_data *)&dp->mon_data;
+	struct hal_sw_mon_ring_entries *sw_mon_entries;
+	struct ath11k_pdev_mon_stats *rx_mon_stats;
+	struct sk_buff *head_msdu, *tail_msdu;
+	u32 rx_bufs_used = 0, mpdu_rx_bufs_used;
+	int quota = 0, ret, status = 0;
+	bool break_dst_ring = false;
+	void *mon_dst_srng;
+	void *ring_entry;
+
+	spin_lock_bh(&pmon->mon_lock);
+
+	if (!ar->monitor_started) {
+		quota = ath11k_dp_rx_process_mon_status(ab, mac_id, napi, budget);
+		spin_unlock_bh(&pmon->mon_lock);
+		return quota;
+	}
+
+	sw_mon_entries = &pmon->sw_mon_entries;
+	rx_mon_stats = &pmon->rx_mon_stats;
+
+	if(pmon->hold_mon_dst_ring) {
+		spin_unlock_bh(&pmon->mon_lock);
+		goto reap_status_ring;
+	}
+
+	mon_dst_srng = &ar->ab->hal.srng_list[dp->rxdma_mon_dst_ring.ring_id];
+	if (!mon_dst_srng) {
+		spin_unlock_bh(&pmon->mon_lock);
+		return status;
+	}
+
+	ath11k_hal_srng_access_begin(ar->ab, mon_dst_srng);
+	while ((ring_entry = ath11k_hal_srng_dst_peek(ar->ab, mon_dst_srng))) {
+		head_msdu = NULL;
+		tail_msdu = NULL;
+
+		mpdu_rx_bufs_used = ath11k_dp_rx_full_mon_mpdu_pop(ar, ring_entry,
+								   &head_msdu,
+								   &tail_msdu,
+								   sw_mon_entries);
+		rx_bufs_used += mpdu_rx_bufs_used;
+
+		if (!sw_mon_entries->end_of_ppdu) {
+			if (head_msdu) {
+				ret = ath11k_dp_rx_full_mon_prepare_mpdu(&ab->dp,
+									 pmon->mon_mpdu,
+									 head_msdu,
+									 tail_msdu);
+				if (ret)
+					break_dst_ring = true;
+			}
+
+			goto next_entry;
+		} else {
+			if (!sw_mon_entries->ppdu_id &&
+			    !sw_mon_entries->mon_status_paddr) {
+				break_dst_ring = true;
+				goto next_entry;
+			}
+		}
+
+		rx_mon_stats->dest_ppdu_done++;
+		pmon->mon_ppdu_status = DP_PPDU_STATUS_START;
+		pmon->mon_status_buf_state = DP_MON_STATUS_LAG;
+		pmon->mon_status_paddr = sw_mon_entries->mon_status_paddr;
+		pmon->hold_mon_dst_ring = true;
+next_entry:
+		ring_entry = ath11k_hal_srng_dst_get_next_entry(ar->ab,
+								mon_dst_srng);
+		if (break_dst_ring)
+			break;
+	}
+
+	ath11k_hal_srng_access_end(ar->ab, mon_dst_srng);
+	spin_unlock_bh(&pmon->mon_lock);
+
+	if (rx_bufs_used) {
+		ath11k_dp_rxbufs_replenish(ar->ab, dp->mac_id,
+					   &dp->rxdma_mon_buf_ring,
+					   rx_bufs_used,
+					   HAL_RX_BUF_RBM_SW3_BM, GFP_ATOMIC);
+	}
+
+reap_status_ring:
+	quota = ath11k_dp_rx_process_full_mon_status_ring(ab, mac_id,
+							  napi, budget);
+
+	return quota;
+}
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_rx.h
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/dp_rx.h
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_rx.h
@@ -99,10 +99,6 @@ int ath11k_dp_htt_tlv_iter(struct ath11k
 			   int (*iter)(struct ath11k_base *ar, u16 tag, u16 len,
 				       const void *ptr, void *data),
 			   void *data);
-int ath11k_dp_rx_process_mon_rings(struct ath11k_base *ab, int mac_id,
-				   struct napi_struct *napi, int budget);
-int ath11k_dp_rx_process_mon_status(struct ath11k_base *ab, int mac_id,
-				    struct napi_struct *napi, int budget);
 int ath11k_dp_rx_mon_status_bufs_replenish(struct ath11k_base *ab, int mac_id,
 					   struct dp_rxdma_ring *rx_ring,
 					   int req_entries,
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_tx.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/dp_tx.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_tx.c
@@ -962,6 +962,46 @@ int ath11k_dp_tx_htt_h2t_ppdu_stats_req(
 	return 0;
 }
 
+int ath11k_dp_tx_htt_rx_full_mon_setup(struct ath11k_base *ab, int mac_id,
+				       bool config)
+{
+	struct htt_rx_full_monitor_mode_cfg_cmd *cmd;
+	struct sk_buff *skb;
+	int ret, len = sizeof(*cmd);
+
+	skb = ath11k_htc_alloc_skb(ab, len);
+	if (!skb)
+		return -ENOMEM;
+
+	skb_put(skb, len);
+	cmd = (struct htt_rx_full_monitor_mode_cfg_cmd*)skb->data;
+	memset(cmd, 0, sizeof(*cmd));
+	cmd->info0 = FIELD_PREP(HTT_RX_FULL_MON_MODE_CFG_CMD_INFO0_MSG_TYPE,
+				HTT_H2T_MSG_TYPE_RX_FULL_MONITOR_MODE);
+
+	cmd->info0 |= FIELD_PREP(HTT_RX_FULL_MON_MODE_CFG_CMD_INFO0_PDEV_ID,
+				   ab->hw_pdev_id_map[mac_id]);
+
+	cmd->cfg = HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_ENABLE |
+		   FIELD_PREP(HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_RELEASE_RING,
+			      HTT_RX_MON_RING_SW);
+	if (config) {
+		cmd->cfg |= HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_ZERO_MPDUS_END |
+			    HTT_RX_FULL_MON_MODE_CFG_CMD_CFG_NON_ZERO_MPDUS_END;
+	}
+
+	ret = ath11k_htc_send(&ab->htc, ab->dp.eid, skb);
+	if (ret)
+		goto err_free;
+
+	return 0;
+
+err_free:
+	dev_kfree_skb_any(skb);
+
+	return ret;
+}
+
 int ath11k_dp_tx_htt_rx_filter_setup(struct ath11k_base *ab, u32 ring_id,
 				     int mac_id, enum hal_ring_type ring_type,
 				     int rx_buf_size,
@@ -1074,10 +1114,20 @@ ath11k_dp_tx_htt_h2t_ext_stats_req(struc
 
 int ath11k_dp_tx_htt_monitor_mode_ring_config(struct ath11k *ar, bool reset)
 {
+	struct ath11k_base *ab = ar->ab;
 	struct ath11k_pdev_dp *dp = &ar->dp;
 	struct htt_rx_ring_tlv_filter tlv_filter = {0};
 	int ret = 0, ring_id = 0;
 
+	if (ab->hw_params.full_monitor_mode) {
+		ret = ath11k_dp_tx_htt_rx_full_mon_setup(ab,
+							 dp->mac_id, !reset);
+		if (ret < 0) {
+			ath11k_err(ab, "failed to setup full monitor %d\n", ret);
+			return ret;
+		}
+	}
+
 	ring_id = dp->rxdma_mon_buf_ring.refill_buf_ring.ring_id;
 
 	if (!reset) {
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_tx.h
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/dp_tx.h
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/dp_tx.h
@@ -237,4 +237,6 @@ int ath11k_dp_tx_htt_rx_filter_setup(str
 enum hal_tcl_encap_type
 ath11k_dp_tx_get_encap_type(struct ath11k_vif *arvif, struct sk_buff *skb);
 
+int ath11k_dp_tx_htt_rx_full_mon_setup(struct ath11k_base *ab, int mac_id,
+				       bool config);
 #endif
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hal_desc.h
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/hal_desc.h
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hal_desc.h
@@ -859,6 +859,25 @@ struct hal_reo_entrance_ring {
  *		this ring has looped around the ring.
  */
 
+#define HAL_SW_MON_RING_INFO0_RXDMA_PUSH_REASON	GENMASK(1, 0)
+#define HAL_SW_MON_RING_INFO0_RXDMA_ERROR_CODE	GENMASK(6, 2)
+#define HAL_SW_MON_RING_INFO0_MPDU_FRAG_NUMBER	GENMASK(10, 7)
+#define HAL_SW_MON_RING_INFO0_FRAMELESS_BAR	BIT(11)
+#define HAL_SW_MON_RING_INFO0_STATUS_BUF_CNT	GENMASK(15, 12)
+#define HAL_SW_MON_RING_INFO0_END_OF_PPDU	BIT(16)
+
+#define HAL_SW_MON_RING_INFO1_PHY_PPDU_ID	GENMASK(15, 0)
+#define HAL_SW_MON_RING_INFO1_RING_ID		GENMASK(27, 20)
+#define HAL_SW_MON_RING_INFO1_LOOPING_COUNT	GENMASK(31, 28)
+
+struct hal_sw_monitor_ring {
+	struct ath11k_buffer_addr buf_addr_info;
+	struct rx_mpdu_desc rx_mpdu_info;
+	struct ath11k_buffer_addr status_buf_addr_info;
+	u32 info0;
+	u32 info1;
+} __packed;
+
 #define HAL_REO_CMD_HDR_INFO0_CMD_NUMBER	GENMASK(15, 0)
 #define HAL_REO_CMD_HDR_INFO0_STATUS_REQUIRED	BIT(16)
 
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hal_rx.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/hal_rx.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hal_rx.c
@@ -1665,3 +1665,50 @@ void ath11k_hal_rx_reo_ent_buf_paddr_get
 
 	*pp_buf_addr = (void *)buf_addr_info;
 }
+
+void
+ath11k_hal_rx_sw_mon_ring_buf_paddr_get(void *rx_desc,
+					struct hal_sw_mon_ring_entries *sw_mon_entries)
+{
+	struct hal_sw_monitor_ring *sw_mon_ring =
+		(struct hal_sw_monitor_ring *)rx_desc;
+	struct ath11k_buffer_addr *buf_addr_info;
+	struct ath11k_buffer_addr *status_buf_addr_info;
+	struct rx_mpdu_desc *rx_mpdu_desc_info_details;
+
+	rx_mpdu_desc_info_details =
+		(struct rx_mpdu_desc *)&sw_mon_ring->rx_mpdu_info;
+
+	sw_mon_entries->msdu_cnt = FIELD_GET(RX_MPDU_DESC_INFO0_MSDU_COUNT,
+			rx_mpdu_desc_info_details->info0);
+
+	buf_addr_info = (struct ath11k_buffer_addr *)&sw_mon_ring->buf_addr_info;
+	status_buf_addr_info =
+		(struct ath11k_buffer_addr *)&sw_mon_ring->status_buf_addr_info;
+
+	sw_mon_entries->mon_dst_paddr = (((u64)FIELD_GET(BUFFER_ADDR_INFO1_ADDR,
+					buf_addr_info->info1)) << 32) |
+					FIELD_GET(BUFFER_ADDR_INFO0_ADDR,
+					buf_addr_info->info0);
+
+	sw_mon_entries->mon_status_paddr =
+			(((u64)FIELD_GET(BUFFER_ADDR_INFO1_ADDR,
+					 status_buf_addr_info->info1)) << 32) |
+				FIELD_GET(BUFFER_ADDR_INFO0_ADDR,
+					  status_buf_addr_info->info0);
+
+	sw_mon_entries->mon_dst_sw_cookie = FIELD_GET(BUFFER_ADDR_INFO1_SW_COOKIE,
+						      buf_addr_info->info1);
+
+	sw_mon_entries->mon_status_sw_cookie = FIELD_GET(BUFFER_ADDR_INFO1_SW_COOKIE,
+							 status_buf_addr_info->info1);
+
+	sw_mon_entries->status_buf_count = FIELD_GET(HAL_SW_MON_RING_INFO0_STATUS_BUF_CNT,
+						     sw_mon_ring->info0);
+
+	sw_mon_entries->dst_buf_addr_info = (void *)buf_addr_info;
+	sw_mon_entries->status_buf_addr_info = (void *)status_buf_addr_info;
+
+	sw_mon_entries->ppdu_id =
+		FIELD_GET(HAL_SW_MON_RING_INFO1_PHY_PPDU_ID, sw_mon_ring->info1);
+}
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hal_rx.h
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/hal_rx.h
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hal_rx.h
@@ -134,6 +134,20 @@ struct hal_rx_user_status {
 	u32 mpdu_err_byte_count;
 };
 
+struct hal_sw_mon_ring_entries {
+	dma_addr_t mon_dst_paddr;
+	dma_addr_t mon_status_paddr;
+	u32 mon_dst_sw_cookie;
+	u32 mon_status_sw_cookie;
+	void *dst_buf_addr_info;
+	void *status_buf_addr_info;
+	u16 ppdu_id;
+	u8 status_buf_count;
+	u8 msdu_cnt;
+	bool end_of_ppdu;
+	bool drop_ppdu;
+};
+
 struct hal_rx_mon_ppdu_info {
 	u32 ppdu_id;
 	u32 last_ppdu_id;
@@ -518,6 +532,9 @@ void ath11k_hal_rx_reo_ent_buf_paddr_get
 					 dma_addr_t *paddr, u32 *sw_cookie,
 					 void **pp_buf_addr_info,
 					 u32 *msdu_cnt);
+void
+ath11k_hal_rx_sw_mon_ring_buf_paddr_get(void *rx_desc,
+					struct hal_sw_mon_ring_entries *sw_mon_entries);
 enum hal_rx_mon_status
 ath11k_hal_rx_parse_mon_status(struct ath11k_base *ab,
 			       struct hal_rx_mon_ppdu_info *ppdu_info,
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hw.c
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/hw.c
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hw.c
@@ -453,6 +453,18 @@ static u32 ath11k_get_reo_dest_remap_con
 	return ring_hash_map;
 }
 
+int ath11k_dp_rx_process_mon_rings_ipq8074(struct ath11k_base *ab, int mac_id,
+					   struct napi_struct *napi, int budget)
+{
+	return ath11k_dp_rx_process_mon_status(ab, mac_id, napi, budget);
+}
+
+int ath11k_dp_rx_process_mon_rings_qcn9000(struct ath11k_base *ab, int mac_id,
+					   struct napi_struct *napi, int budget)
+{
+	return ath11k_dp_rx_process_full_monitor(ab, mac_id, napi, budget);
+}
+
 const struct ath11k_hw_ops ath11k_qca8074_ops = {
 	.get_hw_mac_from_pdev_id = ath11k_qca8074_hw_mac_from_pdev_id,
 	.tx_mesh_enable = ath11k_tx_mesh_enable_qca8074,
@@ -487,6 +499,7 @@ const struct ath11k_hw_ops ath11k_qca807
 	.rx_desc_get_hal_mpdu_peerid = ath11k_rx_desc_get_hal_mpdu_peerid_qca8074,
 	.rx_desc_get_hal_mpdu_len = ath11k_rx_desc_get_hal_mpdu_len_qca8074,
 	.get_reo_dest_remap_config = ath11k_get_reo_dest_remap_config_default,
+	.rx_process_mon_rings = ath11k_dp_rx_process_mon_rings_ipq8074,
 };
 
 const struct ath11k_hw_ops ath11k_qca6018_ops = {
@@ -523,6 +536,7 @@ const struct ath11k_hw_ops ath11k_qca601
 	.rx_desc_get_hal_mpdu_peerid = ath11k_rx_desc_get_hal_mpdu_peerid_qca8074,
         .rx_desc_get_hal_mpdu_len = ath11k_rx_desc_get_hal_mpdu_len_qca8074,
 	.get_reo_dest_remap_config = ath11k_get_reo_dest_remap_config_default,
+	.rx_process_mon_rings = ath11k_dp_rx_process_mon_rings_ipq8074,
 };
 
 const struct ath11k_hw_ops ath11k_qcn9000_ops = {
@@ -559,6 +573,7 @@ const struct ath11k_hw_ops ath11k_qcn900
 	.rx_desc_get_hal_mpdu_peerid = ath11k_rx_desc_get_hal_mpdu_peerid_qcn9000,
 	.rx_desc_get_hal_mpdu_len = ath11k_rx_desc_get_hal_mpdu_len_qcn9000,
 	.get_reo_dest_remap_config = ath11k_get_reo_dest_remap_config_default,
+	.rx_process_mon_rings = ath11k_dp_rx_process_mon_rings_qcn9000,
 };
 
 const struct ath11k_hw_ops ath11k_ipq5018_ops = {
@@ -595,5 +610,6 @@ const struct ath11k_hw_ops ath11k_ipq501
 	.rx_desc_get_hal_mpdu_peerid = ath11k_rx_desc_get_hal_mpdu_peerid_qcn9000,
 	.rx_desc_get_hal_mpdu_len = ath11k_rx_desc_get_hal_mpdu_len_qcn9000,
 	.get_reo_dest_remap_config = ath11k_get_reo_dest_remap_config_ipq5018,
+	.rx_process_mon_rings = ath11k_dp_rx_process_mon_rings_qcn9000,
 };
 
Index: backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hw.h
===================================================================
--- backports-20200908-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/hw.h
+++ backports-20200908-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/hw.h
@@ -134,6 +134,7 @@ struct ath11k_base;
 struct hal_rx_desc;
 struct hal_tcl_data_cmd;
 struct hal_rx_mpdu_info;
+struct napi_struct;
 
 struct ath11k_hw_ops {
 	u8 (*get_hw_mac_from_pdev_id)(struct ath11k_base *ab, int pdev_id);
@@ -172,6 +173,8 @@ struct ath11k_hw_ops {
 	u16 (*rx_desc_get_hal_mpdu_peerid)(struct hal_rx_mpdu_info *mpdu_info);
 	u32 (*rx_desc_get_hal_mpdu_len)(struct hal_rx_mpdu_info *mpdu_info);
 	u32 (*get_reo_dest_remap_config)(void);
+	int (*rx_process_mon_rings)(struct ath11k_base *ab, int mac_id,
+				    struct napi_struct *napi, int budget);
 };
 
 struct ath11k_hw_params {
@@ -192,6 +195,7 @@ struct ath11k_hw_params {
 	u8 spectral_summary_pad_sz;
 	u8 spectral_fft_hdr_len;
 	bool credit_flow;
+	bool full_monitor_mode;
 	u32 hal_desc_sz;
 	u32 m3_addr;
 	u32 spectral_max_fft_bins;
@@ -284,4 +288,11 @@ extern const struct ath11k_hw_ops ath11k
 extern const struct ath11k_hw_ops ath11k_qcn9000_ops;
 extern const struct ath11k_hw_ops ath11k_ipq5018_ops;
 
+int ath11k_dp_rx_process_mon_rings(struct ath11k_base *ab, int mac_id,
+				   struct napi_struct *napi, int budget);
+int ath11k_dp_rx_process_full_monitor(struct ath11k_base *ab, int mac_id,
+				      struct napi_struct *napi, int budget);
+int ath11k_dp_rx_process_mon_status(struct ath11k_base *ab, int mac_id,
+				    struct napi_struct *napi, int budget);
+
 #endif
